File: csharp/exercises/Webshop.Service/OrderService.cs
namespace Webshop.Service;

public interface IOrderService
{
    OrderResult PlaceOrder(string productId, int quantity, string userId);
}

public class OrderService : IOrderService
{
    private readonly IOrderRepository _orderRepository;
    private readonly IProductService _productService;
    private readonly IUserService _userService;
    private readonly INotificationService _notificationService;

    public OrderService(IOrderRepository orderRepository, IProductService productService, IUserService userService, INotificationService notificationService)
    {
        _orderRepository = orderRepository;
        _productService = productService;
        _userService = userService;
        _notificationService = notificationService;
    }

    public OrderResult PlaceOrder(string productId, int quantity, string userId)
    {
        // Verify input arguments
        if (string.IsNullOrEmpty(productId))
            throw new ArgumentNullException(nameof(productId), "Product id must be specified");
        if (quantity <= 0)
            throw new ArgumentOutOfRangeException(nameof(quantity), "Quantity must be greater than zero");
        if (string.IsNullOrEmpty(userId))
            throw new ArgumentNullException(nameof(userId), "User id must be specified");

        // Get product-info with price
        var product = _productService.GetProduct(productId);
        if (product == null)
            return new OrderResult { ResultCode = OrderResultCode.ProductNotFound };

        // Get inventory-info and verify that we have enough items of the specified product
        var inventory = _productService.GetProductInventory(productId);
        if (inventory < quantity)
            return new OrderResult { ResultCode = OrderResultCode.OutOfStock };

        // Handle payment
        var paymentAmount = product.Price * quantity;
        // PaymentService is not yet very testable - we will fix this in the last part of the workshop
        //_paymentService.HandlePayment(userId, paymentAmount);

        // Update inventory
        _productService.SubtractFromProductInventory(productId, quantity);

        // Send notification
        var user = _userService.GetUser(userId);
        _notificationService.SendMail(user.Email, $"Order confirmation: Your order for product {productId} has been succesful.");

        // Create and return order
        var order = _orderRepository.AddOrder(productId, quantity, userId);
        return new OrderResult { Order = order, ResultCode = OrderResultCode.Successful };
    }
}

public class OrderResult
{
    public bool IsSuccess => ResultCode == OrderResultCode.Successful;
    public OrderResultCode ResultCode { get; set; }
    public Order? Order { get; set; }
}

public enum OrderResultCode
{
    Successful = 0,
    ProductNotFound = 1,
    OutOfStock = 2,
    PaymentFailed = 3
}

File: java/exercises/src/main/java/Webshop/Service/Orders/OrderService.java
package Webshop.Service.Orders;

import Webshop.Service.Common.NotificationService;
import Webshop.Service.Products.ProductService;
import Webshop.Service.Users.User;
import Webshop.Service.Users.UserService;
import Webshop.Service.Products.Product;
import Webshop.Service.Orders.OrderResultCode;
import Webshop.Service.Orders.OrderResult;

import java.util.Objects;

interface IOrderService {
    OrderResult placeOrder(String productId, int quantity, String userId);
}

public class OrderService implements IOrderService {
    private final OrderRepository orderRepository;
    private final ProductService productService;
    private final UserService userService;
    private final NotificationService notificationService;

    public OrderService(OrderRepository orderRepository, ProductService productService, UserService userService, NotificationService notificationService) {
        this.orderRepository = Objects.requireNonNull(orderRepository, "Order repository must not be null");
        this.productService = Objects.requireNonNull(productService, "Product service must not be null");
        this.userService = Objects.requireNonNull(userService, "User service must not be null");
        this.notificationService = Objects.requireNonNull(notificationService, "Notification service must not be null");

    }

    @Override
    public OrderResult placeOrder(String productId, int quantity, String userId) {
        // Verify input arguments
        if (productId == null || productId.isEmpty()) {
            throw new IllegalArgumentException("Product id must be specified");
        }
        if (quantity <= 0) {
            throw new IllegalArgumentException("Quantity must be greater than zero");
        }
        if (userId == null || userId.isEmpty()) {
            throw new IllegalArgumentException("User id must be specified");
        }

        // Get product-info with price
        Product product = productService.getProduct(productId);
        if (product == null) {
            return new OrderResult(OrderResultCode.ProductNotFound);
        }

        // Get inventory-info and verify that we have enough items of the specified product
        int inventory = productService.getProductInventory(productId);
        if (inventory < quantity) {
            return new OrderResult(OrderResultCode.OutOfStock);
        }

        // Handle payment
        double paymentAmount = product.getPrice() * quantity;
        // PaymentService is not yet very testable - we will fix this in the last part of the workshop
//        paymentService.handlePayment(userId, paymentAmount);

        // Update inventory
        productService.subtractFromProductInventory(productId, quantity);

        // Send notification
        User user = userService.getUser(userId);
        notificationService.sendMail(user.getEmail(), "Order confirmation: Your order for product " + productId + " has been successful.");

        // Create and return order
        Order order = orderRepository.addOrder(productId, quantity, userId);
        return new OrderResult(OrderResultCode.Successful, order);
    }
}